frontend/lib/resolution/try-catch-analysis.cpp

```diff
@@ -0,0 +1,364 @@
+    // need to keep a stack of try nodes and some meta info
+    std::vector<TryCatchState> tryStack;
+    const Function* currentFn = nullptr;
+    // output
+    // TODO: figure out the output - errors when something doesn't pass?
+    // maybe no output?
+    std::vector<QualifiedType> returnedTypes;
+
+    TryCatchAnalyzer(Context* context,
+                  const AstNode* symbol,
+                  ErrorCheckingMode mode)
+        : context(context),
+          astForErr(symbol),
+          mode(mode) {
+    }
+
+    void checkCatchallIsLast(const Try* node, RV& rv) {
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (hasCatchAll) {
+          context->error(node, "catchall placed before the end of a catch list");
+        }
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+        }
+      }
+    }
+
+    // Returns true if the catch is a catchall statement.
+    bool isCatchAll(const Catch* node, RV& rv) {
+        bool catchAll = false;
+        if (auto err = node->error()) {
+          // we validate that the type of error is a subtype of Error or Error
+          // itself resolving the AST.
+          // So here we just need to check for the base Errors.Error to know if
+          // this is a catchall.
+          const ResolvedExpression& r = rv.byAst(err);
+          auto errType = CompositeType::getErrorType(context);
+          auto dec = ClassTypeDecorator(ClassTypeDecorator::MANAGED_NONNIL);
+          auto manager = AnyOwnedType::get(context);
+          auto ret = ClassType::get(context, errType->basicClassType(),
+                                    manager, dec);
+          auto qt = QualifiedType(QualifiedType::VAR,
+                                  ret->withDecorator(context, dec));
+          auto passResult = canPass(context, r.type(), qt);
+          using CPR=CanPassResult::ConversionKind;
+          if (passResult.passes() && passResult.conversionKind()==CPR::NONE) {
+            catchAll = true;
+          }
+        } else {
+          // this is a catchall since it has no error associated with it
+          catchAll = true;
+        }
+        return catchAll;
+    }
+
+    // Returns true if the catches cover all of the cases.
+     bool catchesExhaustive(const Try* node, RV& rv) {
+      if (node->isTryBang()) {
+        return true;
+      }
+
+      bool hasCatchAll = false;
+      for (auto handler : node->handlers()) {
+        if (isCatchAll(handler, rv)) {
+          hasCatchAll = true;
+          break;
+        }
+      }
+
+      return hasCatchAll;
+    }
+
+    bool canThrow() {
+      return currentFn && currentFn->throws();
+    }
+
+    void enterAst(const Try* ast, RV& rv) {
+        TryCatchState tca;
+        tca.tryNode = ast;
+        tca.catchesAreExhaustive = catchesExhaustive(ast, rv);
+        tca.taskFunctionDepth = 0;
+        tryStack.push_back(tca);
+    }
+
+    void exitAst() {
+      tryStack.pop_back();
+    }
```

Comment:
These are now specifically for `try`s -- maybe call them `enterTry` / `exitTry`?